src/libraries/Common/tests/System/Net/Http/Http3LoopbackStream.cs

```diff
@@ -0,0 +1,326 @@
+                        request.Method = header.Value;
+                        break;
+                    case ":path":
+                        request.Path = header.Value;
+                        break;
+                }
+            }
+
+            return request;
+        }
+
+        public async Task WaitForCancellationAsync(bool ignoreIncomingData = true)
+        {
+            while (true)
+            {
+                (long? frameType, _) = await ReadFrameAsync().ConfigureAwait(false);
+
+                switch (frameType)
+                {
+                    case null:
+                        // end of stream reached.
+                        return;
+                    case DataFrame when ignoreIncomingData == true:
+                        break;
+                    default:
+                        Debug.Fail("Unexpected frame type while waiting for client cancellation.");
+                        throw new Exception();
+                }
+            }
+        }
+
+        public async Task<(long? frameType, byte[] payload)> ReadFrameAsync()
+        {
+            long? frameType = await ReadInteger().ConfigureAwait(false);
+            if (frameType == null) return (null, null);
+
+            long? payloadLength = await ReadInteger().ConfigureAwait(false);
+            if (payloadLength == null) throw new Exception("Unable to read frame; unexpected end of stream.");
+
+            byte[] payload = new byte[checked((int)payloadLength)];
+            int totalBytesRead = 0;
+
+            while (totalBytesRead != payloadLength)
+            {
+                int bytesRead = await _stream.ReadAsync(payload.AsMemory(totalBytesRead)).ConfigureAwait(false);
+                if (bytesRead == 0) throw new Exception("Unable to read frame; unexpected end of stream.");
+
+                totalBytesRead += bytesRead;
+            }
+
+            return (frameType, payload);
+        }
+
+        public async Task<long?> ReadInteger()
+        {
+            byte[] buffer = new byte[MaximumVarIntBytes];
+            int bufferAvailableOffset = -1;
+            int bufferActiveLength = 0;
+
+            long integerValue;
+            int bytesRead;
+
+            do
+            {
+                bytesRead = await _stream.ReadAsync(buffer.AsMemory(++bufferAvailableOffset, 1)).ConfigureAwait(false);
+                if (bytesRead == 0)
+                {
+                    return bufferActiveLength == 0 ? (long?)null : throw new Exception("Unable to read varint; unexpected end of stream.");
+                }
+                Debug.Assert(bytesRead == 1);
+            }
+            while (!TryDecodeHttpInteger(buffer.AsSpan(0, ++bufferActiveLength), out integerValue, out bytesRead));
+
+            Debug.Assert(bytesRead == bufferActiveLength);
+
+            return integerValue;
+        }
+
+        static bool TryDecodeHttpInteger(ReadOnlySpan<byte> buffer, out long value, out int bytesRead)
```

Comment:
Intentional -- our loopback code does not share product code.

Reply:
eek. That seems like an odd decision to me. What are the main reasons for not sharing "product code". Can you do internals visible to?

Reply:
> What are the main reasons for not sharing "product code".

Two implementations of something make it easier to find bugs, and the test implementation often sacrifices perf for flexibility or simplicity.

> Can you do internals visible to?

If we did anything it would be to include the source directly; InternalsVisibleTo is a hard no per various discussions on its goodness.